
<style>
  .notebook-container {
    font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif;
    max-width: 900px;
    margin: 0 auto;
    background-color: #ffffff;
    padding: 20px;
    border-radius: 15px;
    box-shadow: 0 4px 12px rgba(0, 0, 0, 0.1);
  }
  h1 {
    color: #000000;
    background: #FFFACD;
    padding: 18px;
    border-radius: 15px 50px;
    font-family: 'New Timeroman', serif;
    text-align: center;
    margin-top: 30px;
    margin-bottom: 20px;
  }
  h2 {
    color: #000000;
    background: #F0E68C;
    padding: 12px;
    border-radius: 8px;
    font-family: 'New Timeroman', serif;
    margin-top: 25px;
    margin-bottom: 15px;
  }
  p {
    line-height: 1.6;
    color: #333333;
    margin-bottom: 15px;
  }
  ul {
    background: #fafafa;
    padding: 15px 20px;
    border-left: 4px solid #FFD700;
    border-radius: 0 8px 8px 0;
    margin: 15px 0;
  }
  li {
    margin-bottom: 8px;
    color: #444444;
  }
  .highlight {
    background-color: #FFF3E0;
    padding: 2px 6px;
    border-radius: 4px;
  }
</style>

<div class="notebook-container">

<h1 id="iris-classification-analysis" style="color:#000000;background:#FFFACD;padding:18px;border-radius:15px 50px; font-family:newtimeroman;">
  классификация ирисов (IrisClassificationNN.ipynb)
</h1>


##  Подготовка данных с нормализацией


<h2 style="color:#000000;background:#FFFACD;padding:8px;border-radius:8px;font-family:newtimeroman;">
  Построчное объяснение выполнения
</h2>

<h3 style="color:#000000;background:#F0F8FF;padding:6px;border-radius:6px;font-family:newtimeroman;">
  Блок импортов библиотек
</h3>
<p>Импортирует numpy для численных вычислений, pandas для работы с DataFrame, функции sklearn для загрузки данных, разделения выборки, стандартизации и метрик качества. Импортирует tensorflow.keras для создания нейронной сети и matplotlib для построения графиков.</p>

<h3 style="color:#000000;background:#F0F8FF;padding:6px;border-radius:6px;font-family:newtimeroman;">
  Загрузка данных
</h3>
<p><code>iris = load_iris()</code> выполняет загрузку датасета Iris из sklearn.datasets. Функция возвращает объект Bunch, содержащий массив данных iris.data (150×4), целевые метки iris.target (150), названия признаков iris.feature_names и названия классов iris.target_names.</p>

<h3 style="color:#000000;background:#F0F8FF;padding:6px;border-radius:6px;font-family:newtimeroman;">
  Создание DataFrame
</h3>
<p><code>iris_df = pd.DataFrame(iris.data, columns=iris.feature_names)</code> преобразует массив данных в pandas DataFrame, присваивая названия колонок из iris.feature_names. <code>iris_df['target'] = iris.target</code> добавляет столбец с целевыми метками классов (0, 1, 2).</p>

<h3 style="color:#000000;background:#F0F8FF;padding:6px;border-radius:6px;font-family:newtimeroman;">
  Выделение признаков и меток
</h3>
<p><code>X = iris_df[iris.feature_names]</code> извлекает матрицу признаков (150×4). <code>y = iris_df['target']</code> выделяет вектор целевых меток (150).</p>

<h3 style="color:#000000;background:#F0F8FF;padding:6px;border-radius:6px;font-family:newtimeroman;">
  Разделение выборки
</h3>
<p><code>train_test_split(X, y, test_size=0.3, random_state=19)</code> разделяет данные на обучающую (70%) и тестовую (30%) выборки. Параметр test_size=0.3 определяет долю тестовой выборки. random_state=19 обеспечивает воспроизводимость разбиения. Возвращает X_train (105×4), X_test (45×4), y_train (105), y_test (45).</p>

<h3 style="color:#000000;background:#F0F8FF;padding:6px;border-radius:6px;font-family:newtimeroman;">
  Стандартизация данных
</h3>
<p><code>scaler = StandardScaler()</code> создает объект для стандартизации. <code>X_train_scaled = scaler.fit_transform(X_train)</code> вычисляет среднее и стандартное отклонение по обучающей выборке и преобразует её по формуле z=(x-μ)/σ. <code>X_test_scaled = scaler.transform(X_test)</code> применяет те же параметры (среднее и σ из обучающей выборки) к тестовой выборке без повторного пересчета.</p>

<h2 style="color:#000000;background:#FFFACD;padding:8px;border-radius:8px;font-family:newtimeroman;">
  Последовательность выполнения
</h2>
<ol>
  <li>Загружает датасет Iris (150 образцов, 4 признака, 3 класса).</li>
  <li>Создает структурированный DataFrame.</li>
  <li>Разделяет данные на обучающую (105) и тестовую (45) выборки.</li>
  <li>Выполняет стандартизацию обучающей выборки (fit_transform).</li>
  <li>Применяет те же параметры стандартизации к тестовой выборке (transform).</li>
</ol>

<h2 style="color:#000000;background:#FFFACD;padding:8px;border-radius:8px;font-family:newtimeroman;">
  Назначение этапа
</h2>
<p>Подготавливает данные для обучения моделей классификации. Стандартизация обеспечивает одинаковый масштаб признаков, что критически важно для градиентных методов оптимизации в нейронных сетях.</p>


1

In [118]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, classification_report
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.callbacks import EarlyStopping
import matplotlib.pyplot as plt

# Загрузка и подготовка 
iris = load_iris()
iris_df = pd.DataFrame(iris.data, columns=iris.feature_names)
iris_df['target'] = iris.target

# Разбиение с нормализацией 
X = iris_df[iris.feature_names]
y = iris_df['target']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=19)

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


<h2 style="color:#000000;background:#F0F8FF;padding:6px;border-radius:6px;font-family:newtimeroman;">
  Архитектура Sequential модели
</h2>
<p>Создает последовательную нейронную сеть. Поток данных: 4 входа → Dense(16,ReLU) → BatchNorm → Dropout(0.2) → Dense(12,ReLU) → Dropout(0.2) → Dense(8,ReLU) → Dense(3,Softmax).</p>

<h2 style="color:#000000;background:#F0F8FF;padding:6px;border-radius:6px;font-family:newtimeroman;">
  Слои и их функции
</h2>
<ul>
  <li><code>Dense(16, input_shape=(4,), relu)</code> — первый полносвязный слой, 16 нейронов, ReLU активация</li>
  <li><code>BatchNormalization()</code> — нормализация активаций по батчу, стабилизирует градиенты</li>
  <li><code>Dropout(0.2)</code> — случайное отключение 20% нейронов, предотвращает переобучение</li>
  <li><code>Dense(3, softmax)</code> — выходной слой, 3 нейрона для 3 классов, softmax для вероятностей</li>
</ul>

<h2 style="color:#000000;background:#F0F8FF;padding:6px;border-radius:6px;font-family:newtimeroman;">
  Компиляция модели
</h2>
<p><code>compile()</code> настраивает обучение: Adam оптимизатор (адаптивный градиентный спуск), sparse_categorical_crossentropy (потери для целочисленных меток), accuracy (метрика качества).</p>

<h2 style="color:#000000;background:#F0F8FF;padding:6px;border-radius:6px;font-family:newtimeroman;">
  EarlyStopping callback
</h2>
<p>Останавливает обучение при отсутствии улучшения val_loss в течение 15 эпох. restore_best_weights=True восстанавливает лучшие веса.</p>

<h2 style="color:#000000;background:#F0F8FF;padding:6px;border-radius:6px;font-family:newtimeroman;">
  Обучение модели
</h2>
<p><code>fit()</code> запускает обучение: 200 эпох максимум, батчи по 16 образцов, валидация на тестовой выборке, применение EarlyStopping.</p>

<h2 style="color:#000000;background:#FFFACD;padding:8px;border-radius:8px;font-family:newtimeroman;">
  Последовательность выполнения
</h2>
<ol>
  <li>Создает Sequential модель с 7 слоями</li>
  <li>Компилирует с Adam + sparse_categorical_crossentropy</li>
  <li>Настраивает EarlyStopping по val_loss</li>
  <li>Обучает 200 эпох с батчами по 16 образцов</li>
</ol>

<h2 style="color:#000000;background:#FFFACD;padding:8px;border-radius:8px;font-family:newtimeroman;">
  Результат этапа
</h2>
<p>Получает обученную нейронную сеть history объект с метриками по эпохам. Модель готова к предсказаниям и оценке качества.</p>


2

In [119]:
nn_improved = tf.keras.Sequential([
    tf.keras.layers.Input(shape=(4,)),
    tf.keras.layers.Dense(16, activation='relu', name='input'),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(12, activation='relu', name='hidden1'),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(8, activation='relu', name='hidden2'),
    tf.keras.layers.Dense(3, activation='softmax', name='output')
])


nn_improved.compile(optimizer='adam', 
                   loss='sparse_categorical_crossentropy', 
                   metrics=['accuracy'])

early_stop = EarlyStopping(monitor='val_loss', patience=15, restore_best_weights=True)

history = nn_improved.fit(X_train_scaled, y_train, epochs=200, batch_size=16,
                         validation_data=(X_test_scaled, y_test),
                         callbacks=[early_stop], verbose=0)



<h2 style="color:#000000;background:#F0F8FF;padding:6px;border-radius:6px;font-family:newtimeroman;">
  Создание трех классических моделей
</h2>

<h3 style="color:#000000;background:#F0F8FF;padding:4px;border-radius:4px;font-family:newtimeroman;">
  RandomForestClassifier
</h3>
<p>Ансамбль 100 решающих деревьев. n_estimators=100 — количество деревьев. Бэггинг + случайные подвыборки признаков.</p>

<h3 style="color:#000000;background:#F0F8FF;padding:4px;border-radius:4px;font-family:newtimeroman;">
  SVC (RBF ядро)
</h3>
<p>Метод опорных векторов с радиально-базисным ядром. kernel='rbf' — нелинейная разделяющая гиперповерхность. probability=True — выводит вероятности.</p>

<h3 style="color:#000000;background:#F0F8FF;padding:4px;border-radius:4px;font-family:newtimeroman;">
  LogisticRegression
</h3>
<p>Многоклассовая логистическая регрессия</p>

<h2 style="color:#000000;background:#FFFACD;padding:8px;border-radius:8px;font-family:newtimeroman;">
  Словарь моделей
</h2>
<p>Словарь models содержит названия и объекты моделей для удобного циклического обучения и сравнения.</p>

<h2 style="color:#000000;background:#FFFACD;padding:8px;border-radius:8px;font-family:newtimeroman;">
  Последовательность
</h2>
<ol>
  <li>Импортирует три классификатора sklearn</li>
  <li>Создает экземпляры с параметрами по умолчанию + random_state=19</li>
  <li>Собирает в словарь для унифицированного обращения</li>
</ol>

<h2 style="color:#000000;background:#FFFACD;padding:8px;border-radius:8px;font-family:newtimeroman;">
  Назначение
</h2>
<p>Подготавливает базовые модели машинного обучения для сравнения с нейронной сетью на единой тестовой выборке.</p>


In [120]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression

# 3 классические модели
rf_model = RandomForestClassifier(n_estimators=100, random_state=19)
svm_model = SVC(kernel='rbf', probability=True, random_state=19, C=1.0)
lr_model = LogisticRegression(random_state=19)  

models = {
    'Random Forest': rf_model,
    'SVM RBF': svm_model,
    'Logistic Regression': lr_model
}



<h2 style="color:#000000;background:#F0F8FF;padding:6px;border-radius:6px;font-family:newtimeroman;">
  Выполняемые операции
</h2>
<ol>
  <li>Создает пустые словари results и trained_models</li>
  <li>Для нейронной сети: predict() → argmax() → accuracy_score()</li>
  <li>Цикл по классическим моделям: fit() → predict() → accuracy_score()</li>
  <li>Сохраняет точность и предсказания для каждой модели</li>
</ol>

<h2 style="color:#000000;background:#F0F8FF;padding:6px;border-radius:6px;font-family:newtimeroman;">
  Ключевые моменты
</h2>
<ul>
  <li><code>np.argmax(..., axis=1)</code> преобразует вероятности [0.9,0.1,0.0] → класс 0</li>
  <li>Все модели тестируются на одних X_test_scaled, y_test</li>
  <li><code>accuracy_score()</code> вычисляет долю правильных предсказаний</li>
</ul>

<h2 style="color:#000000;background:#FFFACD;padding:8px;border-radius:8px;font-family:newtimeroman;">
  Результат
</h2>
<p>Получает словарь results с точностью 4 моделей и trained_models с обученными классификаторами sklearn.</p>


3

In [121]:
results = {}
trained_models = {}

# Нейронная сеть (получаем)
y_pred_nn = np.argmax(nn_improved.predict(X_test_scaled), axis=1)
results['Neural Network (Improved)'] = {
    'accuracy': accuracy_score(y_test, y_pred_nn),
    'predictions': y_pred_nn
}

# Классические модели
for name, model in models.items():
    model.fit(X_train_scaled, y_train)
    trained_models[name] = model
    y_pred = model.predict(X_test_scaled)
    results[name] = {
        'accuracy': accuracy_score(y_test, y_pred),
        'predictions': y_pred
    }


2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 55ms/step



<h2 style="color:#000000;background:#F0F8FF;padding:6px;border-radius:6px;font-family:newtimeroman;">
  Выполняемые операции
</h2>
<ol>
  <li>Импортирует VotingClassifier из sklearn.ensemble</li>
  <li>Создает ансамбль из RandomForest и SVM</li>
  <li>Обучает ансамбль на X_train_scaled</li>
  <li>Получает предсказания на тестовой выборке</li>
  <li>Вычисляет точность и сохраняет в results</li>
</ol>

<h2 style="color:#000000;background:#F0F8FF;padding:6px;border-radius:6px;font-family:newtimeroman;">
  Параметры VotingClassifier
</h2>
<ul>
  <li><code>estimators=[('rf', rf_model), ('svm', svm_model)]</code> — две базовые модели</li>
  <li><code>voting='soft'</code> — усреднение вероятностей (лучше чем голосование по классам)</li>
</ul>

<h2 style="color:#000000;background:#FFFACD;padding:8px;border-radius:8px;font-family:newtimeroman;">
  Принцип работы
</h2>
<p>Каждая модель выдает вероятности для 3 классов. Ансамбль усредняет вероятности RF и SVM, выбирает класс с максимальной средней вероятностью.</p>

<h2 style="color:#000000;background:#FFFACD;padding:8px;border-radius:8px;font-family:newtimeroman;">
  Результат
</h2>
<p>Получает 5-ю модель в сравнении — ансамбль RandomForest+SVM. Добавляет точность в словарь results.</p>


In [122]:
from sklearn.ensemble import VotingClassifier

# Soft voting ensemble (лучшие модели)
ensemble = VotingClassifier(
    estimators=[('rf', rf_model), ('svm', svm_model)],
    voting='soft'
)
ensemble.fit(X_train_scaled, y_train)
y_pred_ens = ensemble.predict(X_test_scaled)
results['Ensemble'] = {
    'accuracy': accuracy_score(y_test, y_pred_ens),
    'predictions': y_pred_ens
}

 <h2 style="color:#000000;background:#F0F8FF;padding:6px;border-radius:6px;font-family:newtimeroman;"> Выполняемые операции </h2> <ol> <li>Выводит заголовок сравнения</li> <li>Создает DataFrame из словаря results</li> <li>Преобразует accuracy (0.98) в проценты (98.0%)</li> <li>Сортирует по убыванию точности</li> <li>Выводит таблицу</li> </ol> <h2 style="color:#000000;background:#F0F8FF;padding:6px;border-radius:6px;font-family:newtimeroman;"> Ключевые элементы </h2> <ul> <li><code>list(results.keys())</code> — названия моделей</li> <li><code>v['accuracy']*100:.1f</code> — точность в % с 1 знаком</li> <li><code>sort_values(ascending=False)</code> — от лучшей к худшей</li> </ul> <h2 style="color:#000000;background:#FFFACD;padding:8px;border-radius:8px;font-family:newtimeroman;"> Результат </h2> <p>Получает отсортированную таблицу сравнения точности 5 моделей на тестовой выборке.</p>

In [123]:
# Вывод результатов
print("\n=== СРАВНЕНИЕ МОДЕЛЕЙ ===")
comparison_df = pd.DataFrame({
    'Модель': list(results.keys()),
    'Accuracy (%)': [f"{v['accuracy']*100:.1f}" for v in results.values()]
}).sort_values('Accuracy (%)', ascending=False)
print(comparison_df)



=== СРАВНЕНИЕ МОДЕЛЕЙ ===
                      Модель Accuracy (%)
1              Random Forest         93.3
2                    SVM RBF         93.3
3        Logistic Regression         93.3
4                   Ensemble         93.3
0  Neural Network (Improved)        100.0


<h1 id="final-results" align="center" style="color:#000000;background:#FFFACD;padding:18px;border-radius:15px 50px; font-family:newtimeroman;"> Итоговое сравнение моделей классификации Iris </h1> <h1 id="top-models" style="color:#000000;background:#FFFACD;padding:8px;border-radius:8px;font-family:newtimeroman;"> Результаты моделей </h1> <ul> <li><b>Нейронная сеть (улучшенная)</b> — 97.8%. Dense(16→12→8) + BatchNorm + Dropout + Adam</li> <li><b>Random Forest</b> — 93.3%. Ансамбль 100 деревьев, бэггинг</li> <li><b>SVM RBF + Логистическая регрессия</b> — 93.3%. Нелинейное ядро и линейная модель</li> </ul> <h1 id="key-conclusion" style="color:#000000;background:#FFFACD;padding:8px;border-radius:8px;font-family:newtimeroman;"> Основные выводы </h1> <ul> <li>Нейронная сеть лидирует за счет глубокой архитектуры и нормализации данных</li> <li>Random Forest показывает стабильный результат ансамблевого метода</li> <li>Все модели демонстрируют высокое качество (>93%) на датасете Iris</li> </ul> <h2 style="color:#000000;background:#FFFACD;padding:8px;border-radius:8px;font-family:newtimeroman;"> Лучший выбор
</h2> <p><b>Нейронная сеть</b> — 97.8% точности</p>

Нейронная модель лучше справляется с классификацией ирисов и лучше находит нелинейные признаки; 
Ансамбль получился слабым из-за слабых моделей;
Так же улучшенная модель в отличии от преведущей может выдавать результат в 100%-97,8%
